etl/grapher_import.py

```diff
@@ -67,62 +67,80 @@ def upsert_dataset(dataset: catalog.Dataset, namespace: str, sources: List[catal
             " join multiple sources"
         )
 
+    engine = gm.get_engine()
+
     # This function creates the dataset table row, a namespace row
     # and the sources table row(s). There is a bit of an open question if we should
     # map one dataset with N tables to one namespace and N datasets in
     # mysql or if we should just flatten it into one dataset?
-    with open_db() as db:
+    with Session(engine) as session:
         log.info("upsert_dataset.verify_namespace", namespace=namespace)
-        ns = db.fetch_one_or_none("SELECT * from namespaces where name=%s", [namespace])
-        if ns is None:
-            db.upsert_namespace(namespace, "")
-        else:
-            _, ns_name, _, ns_is_archived = ns
-            if ns_is_archived:
-                log.warning("upsert_dataset.namespace_is_archived", namespace=ns_name)
+        ns = gm.Namespace(name=namespace, description="")
+        ns = ns.upsert(session)
+        if ns.isArchived:
+            log.warning("upsert_dataset.namespace_is_archived", namespace=ns.name)
 
         log.info(
             "upsert_dataset.upsert_dataset.start",
             short_name=dataset.metadata.short_name,
         )
-        dataset_id = db.upsert_dataset(
-            dataset.metadata.short_name,
-            dataset.metadata.title,
-            dataset.metadata.version,
-            namespace,
-            int(cast(str, config.GRAPHER_USER_ID)),
+        user_id = int(cast(str, config.GRAPHER_USER_ID))
+        ds = gm.Dataset(
+            shortName=dataset.metadata.short_name,
+            name=dataset.metadata.title,
+            version=dataset.metadata.version,
+            namespace=namespace,
+            metadataEditedByUserId=user_id,
+            dataEditedByUserId=user_id,
+            createdByUserId=user_id,
             description=dataset.metadata.description or "",
         )
+        ds = ds.upsert(session)
+        assert ds.id
+        if ds.isArchived:
+            log.warning(
+                "upsert_dataset.dataset_is_archived",
+                id=ds.id,
+                short_name=ds.shortName,
+            )
+
         log.info(
             "upsert_dataset.upsert_dataset.end",
             short_name=dataset.metadata.short_name,
-            id=dataset_id,
+            id=ds.id,
         )
 
         source_ids: Dict[str, int] = dict()
         for source in sources:
-            source_ids[source.name] = _upsert_source_to_db(db, source, dataset_id)
+            assert source.name
+            source_ids[source.name] = _upsert_source_to_db(session, source, ds.id)
```

Comment:
Consider moving the `_upsert_source_to_db` logic to the grapher model:

```python
source_ids[source.name] = gm.Source.from_catalog_source(source, ds.id).upsert(session)
```

Reply:
Way nicer. I had to keep there some boilerplate code with thread locks (which will go away soon), but it's still much better.